source/common/common/logger.cc

```diff
@@ -55,29 +58,50 @@ void DelegatingLogSink::set_formatter(std::unique_ptr<spdlog::formatter> formatt
 
 void DelegatingLogSink::log(const spdlog::details::log_msg& msg) {
   absl::ReleasableMutexLock lock(&format_mutex_);
-  if (!formatter_) {
-    lock.Release();
-    sink_->log(absl::string_view(msg.payload.data(), msg.payload.size()));
-    return;
-  }
+  absl::string_view msg_view = absl::string_view(msg.payload.data(), msg.payload.size());
 
+  // This memory buffer must exist in the scope of the entire function,
+  // otherwise the string_view will refer to memory that is already free.
   fmt::memory_buffer formatted;
-  formatter_->format(msg, formatted);
+  if (formatter_) {
+    formatter_->format(msg, formatted);
+    msg_view = absl::string_view(formatted.data(), formatted.size());
+  }
   lock.Release();
-  sink_->log(absl::string_view(formatted.data(), formatted.size()));
+
+  if (should_escape_) {
+    // Split the actual log message from the trailing whitespace.
+    auto eol_it = std::find_if_not(msg_view.rbegin(), msg_view.rend(), absl::ascii_isspace);
+    auto msg_leading_view = msg_view.substr(0, msg_view.rend() - eol_it);
+    auto msg_whitespace_view =
+        msg_view.substr(msg_view.rend() - eol_it, eol_it - msg_view.rbegin());
+
+    // Escape the log message, but keep the whitespace unescaped.
+    sink_->log(absl::CEscape(msg_leading_view));
+    sink_->log(msg_whitespace_view);
```

Comment:
do you want to be issuing two separate log calls for the non-whitespace part of the log message and the log message? Or should you combine them in memory and call log once?

Reply:
Done, combined in memory and logged once.